### DISCUSSION TITLE
OAuth Not populating authStore

### DISCUSSION THREAD

**Discussion by timmyjl12** [pocketbase/pocketbase#903] [Q&A]
I am using the default code for Sveltekit in the docs. I am trying to add a Google Login. The issue is, it looks like the authStore is not being populated with the information.

CODE
```
/** @type {import('./$types').PageServerLoad} */
export const load = ({locals, url}) => {
// other information like url.searchParams.get("code")
    locals.pocketbase.users.authViaOAuth2(
      locals.gAuthName,
       url.searchParams.get("code"),
      locals.codeVerifier,
      locals.redirectURL
      ).then((userData) => {
        token = userData.token
        user = userData.user
        // locals.pocketbase.authStore.save(token, user) // Does not work
    }).catch((err) => {
        console.log(err)
    })
 console.log(locals.pocketbase.authStore.model) //blank
}
```

This code will not populate the authStore and comes up blank. Is there something I am missing? Thank you!

**Comment by ganigeorgiev** [comment]
@timmyjl12 I'm not able to reproduce the issue. Are you sure that the OAuth2 authentication completes successfully?

Also, you should use the getter `client.authStore.model` (`baseModel` is [marked "protected"](https://github.com/pocketbase/js-sdk/blob/master/src/stores/BaseAuthStore.ts#L16) and shouldn't be used in non `BaseAuthStore` derived classes).

You could find an example OAuth2 integration in https://pocketbase.io/docs/manage-users/#auth-via-oauth2

**Reply by timmyjl12** [reply]
Hi @ganigeorgiev ! The user data is returned from google successfully and the user is saved in the user model. It's just authStore is not being populated which causes event.locals to not have the user information.

Also, good catch on the baseModel. I forgot to change it back to model in my testing.

**Reply by ganigeorgiev** [reply]
Looking at your code sample, you have both `client` and `locals.pocketbase` instances. Does this mean that you are making the `client.users.authViaOAuth2` call in the browser or it is in some server-side action?

Could you provide a minimal repo where the issue could be seen/reproduced?

**Reply by timmyjl12** [reply]
Hi @ganigeorgiev, I updated my code. My apologies, I have been racking my brain as to figure out why and my code got very sloppy.

I am relying on locals.pocketbase as shown in the docs example. I am utilizing the redirect from google and using that in a load function in my +page.server.js

**Reply by ganigeorgiev** [reply]
I'm still failing to understand what actually the code sample is trying to do. 

What do you mean by the commented line `// locals.pocketbase.authStore.save(token, user) // Does not work`. You don't have to manually save the store. If you console.log the `token` and `user` could you confirm that they are valid?

Also, your last `console.log` may not work because you are not awaiting the `authViaOAuth2` promise. You have to either use `async/await` or move the call inside the `then()` callback.

**Reply by timmyjl12** [reply]
Hi @ganigeorgiev, I'm using the sveltekit load function to get my redirect from google. From there I am using the method locals.pocketbase.users.authViaOAuth2. This all works great, except the model does not seem to be saving. The token and user are valid with console.log.

Yea I originally had it all async await as well. I'm going to change it back to that.

**Reply by ganigeorgiev** [reply]
@timmyjl12 Both `locals.pocketbase.authStore.token` and `locals.pocketbase.authStore.model` are properly set for me when used in a server-side action/endpoint 

The issue is somewhere else, but I'm not sure how to reproduce it. The code sample above needs a little more context. If possible, please provide a minimal reproducible repo with your SvelteKit code and I'll investigate it tomorrow.

**Reply by timmyjl12** [reply]
Hi @ganigeorgiev, I published a quick demo here.

https://github.com/timmyjl12/svelte_pocketbase_tim.git


I can get around the issue by using a redirect in a post. It's bizarre, but it works. I was going to just issue the auth method in the post, but a simple redirect works.

**Comment by hmajid2301** [comment]
Just to add I found something similar, except the authStore would be loaded however when I went away from the page the authStore would be empty i.e. my `hook.server.ts` would be unable to load the authStore. This seemed to be caused by a `throw redirect(303, '/');`. When I comment this line out this seems to all work

```ts
export const GET: RequestHandler = async ({ locals, url, cookies }: RequestEvent) => {
       // Other Code 
       
	try {
		await locals.pb
			?.collection('users')
			.authWithOAuth2(provider.name, code || '', provider.codeVerifier, redirectURL);
	} catch (err) {
		console.log('Error logging in with 0Auth user', err);
	}

	return new Response('123456');
	// throw redirect(303, '/');
};

```

EDIT: Even a redirect response causes issues `return new Response('Redirect', { status: 303, headers: { Location: '/' } });`

Full Code Example: https://gitlab.com/banter-bus/bookmarkey/gui/-/blob/3254b7456a7f93b2b3836371fadb9ab2bbc8f06c/src/routes/account/callback/+server.ts#L38

Am I doing something I am not supposed to ? 

Thanks

**Reply by ganigeorgiev** [reply]
@hmajid2301 Depending how the action is resolved, it is possible that the `throw` could prevent your global hook to reach this line https://gitlab.com/banter-bus/bookmarkey/gui/-/blob/3254b7456a7f93b2b3836371fadb9ab2bbc8f06c/src/hooks.server.ts#L18, which updates the response cookies.

Another reason why the AuthStore may not be set is probably that browsers ignore in some situations the cookies on 303 redirects when samesite=strict is used (both firefox and chrome had issues with this in the past - https://bugzilla.mozilla.org/show_bug.cgi?id=1465402, https://bugs.chromium.org/p/chromium/issues/detail?id=696204).

> Am I doing something I am not supposed to ?

If try/catching the thrown response error doesn't work, you could try temporary changing the exportToCookie to `event.locals.pb.authStore.exportToCookie({ sameSite: 'Lax', secure: isProd })` and see if that will change anything (but keep in mind that in some situation `Lax` may not be sufficient against CSRF attacks; for more info check https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite#lax).

One more thing, in your [hook handler](https://gitlab.com/banter-bus/bookmarkey/gui/-/blob/3254b7456a7f93b2b3836371fadb9ab2bbc8f06c/src/hooks.server.ts), it is recommended to call `refresh` after loading the cookie to verify it server-side, aka. before your `event.locals.pb.authStore.isValid` you can add:

```js
try {
    // get an up-to-date auth store state by veryfing and refreshing the loaded auth model (if any)
    event.locals.pb.authStore.isValid && await event.locals.pb.collection('users').authRefresh();
} catch (_) {
    // clear the auth store on failed refresh
    event.locals.pb.authStore.clear();
}
```
(_if you haven't already, you can also check the [ssr examples](https://github.com/pocketbase/js-sdk#ssr-integration)_)

**Reply by hmajid2301** [reply]
`Lax` seems to have done it and it is now working.

**Reply by jramiroz98** [reply]
Lax worked for me too!
But I'm concerned about the security regarding CSRF attacks
